### PR TITLE
Super Resolution

### PR DESCRIPTION
**Pull request by sushi20**
# Super Resolution

## Use cases
The Super Resolution machine learning model sharpens and upscales the input image to refine the details and improve quality.

## Description
Super Resolution uses efficient  [Sub-pixel convolutional layer](https://arxiv.org/abs/1609.05158) described for increasing spatial resolution within network tasks. By increasing pixel count, images are then clarified, sharpened, and upscaled without losing the input image’s content and characteristics. 

## Model

 |Model      |Download  |Checksum|Download (with sample test data)| ONNX version | Opset Version
|-------------|:--------------|:--------------|:--------------|:--------------| :------------|
|Super_Resolution|    [240 KB](super_resolution.onnx)  |[MD5](super_resolution-md5.txt)  |  [7.6 MB](super_resolution_test_image.gz) |  1.5.0  | 10

### Input 
Image input sizes are dynamic. The inference was done using jpg image. 

### Preprocessing
Images are resized into (224x224). The image format is changed into YCbCr with color components: greyscale ‘Y’, blue-difference  ‘Cb’, and red-difference ‘Cr’. Once the greyscale Y component is extracted, it is then passed through the super resolution model and upscaled. 
  
    from PIL import Image
    from resizeimage import resizeimage
    import numpy as np
    orig_img = Image.open('IMAGE_FILE_PATH')
    img = resizeimage.resize_cover(orig_img, [224,224], validate=False)
    img_ycbcr = img.convert('YCbCr')
    img_y_0, img_cb, img_cr = img_ycbcr.split()
    img_ndarray = np.asarray(img_y_0)
    img_4 = np.expand_dims(np.expand_dims(img_ndarray, axis=0), axis=0)
    img_5 = img_4.astype(np.float32) / 255.0
    img_5
  

### Output
The model outputs a multidimensional array of pixels that are upscaled. Output shape is [batch_size,1,672,672]. The second dimension is one because only the (Y) intensity channel was passed into the super resolution model and upscaled. 

### Postprocessing
Postprocessing involves converting the array of pixels into an image that is scaled to a higher resolution. The color channels (Cb, Cr) are also scaled to a higher resolution using bicubic interpolation. Then the color channels are combined and converted back to RGB format, producing the final output image. 

    final_img = Image.merge(
    "YCbCr", [
        img_out_y,
        img_cb.resize(img_out_y.size, Image.BICUBIC),
        img_cr.resize(img_out_y.size, Image.BICUBIC),
    ]).convert("RGB")
    plt.imshow(final_img)

## Dataset
This model is trained on the [BSD300 Dataset](https://github.com/pytorch/examples/tree/master/super_resolution), using crops from the 200 training images. 

## Training
View the  [training notebook](https://github.com/pytorch/examples/tree/master/super_resolution) to understand details for parameters and network for Super Resolution.

### CODE REVIEW

File: vision/manipulation/super_resolution/README.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+|Super_Resolution v1|    [240 KB](super_resolution_tutorial.onnx)  |[MD5](super_resolution-md5.txt)  |  [7.6 MB](super_resolution_test_image.gz) |  1.5.0  |
+
+## Inference
+
+
+### Input 
+Image file can be jpg, png, and jpeg and its input sizes are dynamic. The inference was done using jpg image.
+
+### Preprocessing
+Images are resized into (224x224). The image is then split into ‘YCbCr’ color components: greyscale ‘Y’, blue-difference  ‘Cb’, and red-difference ‘Cr’. Once the greyscale Y component is extracted, it is then converted to tensor and used as the input image.
```
Comment (prasanthpul) [inline vision/manipulation/super_resolution/README.md]: Why is the size restricted to 224x224? Is it because of the pytorch tracing export?

File: vision/manipulation/super_resolution/README.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    img_ycbcr = img.convert('YCbCr')
+    img_y_0, img_cb, img_cr = img_ycbcr.split()
+    img_ndarray = np.asarray(img_y_0)
+    img_4 = np.expand_dims(np.expand_dims(img_ndarray, axis=0), axis=0)
+    img_5 = img_4.astype(np.float32) / 255.0
+    img_5
+  
+
+### Output
+The model outputs a multidimensional array of pixels that are upscaled. Output shape is [batch_size,1,672,672].
```
Comment (prasanthpul) [inline vision/manipulation/super_resolution/README.md]: the second dimension is the # of channels (colors). 1 means it's monochromatic. Is that right?

File: vision/manipulation/super_resolution/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    "YCbCr", [
+        img_out_y,
+        img_cb.resize(img_out_y.size, Image.BICUBIC),
+        img_cr.resize(img_out_y.size, Image.BICUBIC),
+    ]).convert("RGB")
+    plt.imshow(final_img)
+
+
+## Dataset
+This example trains a super-resolution network on the [BSD300 Dataset](https://github.com/pytorch/examples/tree/master/super_resolution) , using crops from the 200 training images, and evaluating on crops of the 100 test images.
```
Comment (prasanthpul) [inline vision/manipulation/super_resolution/README.md]: "This example trains a super-resolution network" should be "This model has been trained"

Comment (prasanthpul) [inline vision/manipulation/super_resolution/README.md]: How was it evaluated on the 100 test images? Is there a score?

File: vision/manipulation/super_resolution/README.md (hunk 2 above)
Comment (sushi20) [reply vision/manipulation/super_resolution/README.md]: Yea. As the model is being run, it is monochromatic. During the post processing of the image, the image is reverted back to color.

File: vision/manipulation/super_resolution/README.md (hunk 1 above)
Comment (sushi20) [reply vision/manipulation/super_resolution/README.md]: Yea, its due to the tracing export.

File: vision/manipulation/super_resolution/README.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ |Model      |Download  |Checksum|Download (with sample test data)| ONNX version |
+|-------------|:--------------|:--------------|:--------------|:--------------|
+|Super_Resolution v1|    [240 KB](super_resolution_tutorial.onnx)  |[MD5](super_resolution-md5.txt)  |  [7.6 MB](super_resolution_test_image.gz) |  1.5.0  |
+
+## Inference
+
+
+### Input 
+Image file can be jpg, png, and jpeg and its input sizes are dynamic. The inference was done using jpg image.
+
```
Comment (spandantiwari) [inline vision/manipulation/super_resolution/README.md]: It maybe useful to add a paragraph up here to explain at a high-level how super resolution output is obtained. We should clearly mention that the image format is changed to YCbCr and then the intensity channel (Y) is the only one that is passed through the superresolution model to get upscaled intensity image. The color channels undergo a simple bicubic interpolation based upscaling and then as the last steps they are all combined and converted back to RGB format.

File: vision/manipulation/super_resolution/README.md (hunk 2 above)
Comment (spandantiwari) [inline vision/manipulation/super_resolution/README.md]: Good place to point out that second dimension is 1 because we only upscaled the intensity channel (Y) using this model.

File: vision/manipulation/super_resolution/README.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    img_4 = np.expand_dims(np.expand_dims(img_ndarray, axis=0), axis=0)
+    img_5 = img_4.astype(np.float32) / 255.0
+    img_5
+  
+
+### Output
+The model outputs a multidimensional array of pixels that are upscaled. Output shape is [batch_size,1,672,672]. 
+
+### Postprocessing
+Postprocessing involves converting the array of pixels into an image that is scaled to a higher resolution. The ‘YCbCr’ colors are then merged and reconstructed into the final output image.
```
Comment (spandantiwari) [inline vision/manipulation/super_resolution/README.md]: I think we can be clearer here. Something like;
"The color channels (Cb, Cr) are scaled to a higher resolution using bicubic interpolation. Then the output of the model together with upscaled color channels are combined and converted back to RGB format.

File: vision/manipulation/super_resolution/README.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+ |Model      |Download  |Checksum|Download (with sample test data)| ONNX version |
+|-------------|:--------------|:--------------|:--------------|:--------------|
+|Super_Resolution v1|    [240 KB](super_resolution_tutorial.onnx)  |[MD5](super_resolution-md5.txt)  |  [7.6 MB](super_resolution_test_image.gz) |  1.5.0  |
+
+## Inference
+
+
+### Input 
+Image file can be jpg, png, and jpeg and its input sizes are dynamic. The inference was done using jpg image.
```
Comment (ebarsoum) [inline vision/manipulation/super_resolution/README.md]: Why only those 3? Isn't any format supported by Pillow library should work?

File: vision/manipulation/super_resolution/README.md (hunk 4 above)
Comment (sushi20) [reply vision/manipulation/super_resolution/README.md]: Is it better to add this piece of information to the preprocessing stage rather than the input stage?

File: vision/super_resolution/super_resolution_cnn_2016/README.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    "YCbCr", [
+        img_out_y,
+        img_cb.resize(img_out_y.size, Image.BICUBIC),
+        img_cr.resize(img_out_y.size, Image.BICUBIC),
+    ]).convert("RGB")
+    plt.imshow(final_img)
+
+
+## Dataset
+This model is trained on the [BSD300 Dataset](https://github.com/pytorch/examples/tree/master/super_resolution) , using crops from the 200 training images.
```
Comment (prasanthpul) [inline vision/super_resolution/super_resolution_cnn_2016/README.md]: nit: extra space before comma

File: vision/super_resolution/super_resolution_cnn_2016/README.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
+# SuperResolution
```
Comment (prasanthpul) [inline vision/super_resolution/super_resolution_cnn_2016/README.md]: nit: space between Super and Resolution

Comment (prasanthpul) [inline vision/super_resolution/super_resolution_cnn_2016/README.md]: The folder name for these files should be sub_pixel_cnn_2016.
"super_resolution" is already part of the parent folder name

File: vision/super_resolution/super_resolution_cnn_2016/README.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+The Super Resolution machine learning model sharpens and upscales the input image to refine the details and improve quality.
+
+## Description
+Super Resolution uses efficient  [Sub-pixel convolutional layer](https://arxiv.org/abs/1609.05158) described for increasing spatial resolution within network tasks. By increasing pixel count, images are then clarified, sharpened, and upscaled without losing the input image’s content and characteristics. 
+
+## Model
+
+ |Model      |Download  |Checksum|Download (with sample test data)| ONNX version | Opset Version
+|-------------|:--------------|:--------------|:--------------|:--------------| :------------|
+|Super_Resolution v1|    [240 KB](super_resolution_tutorial.onnx)  |[MD5](super_resolution-md5.txt)  |  [7.6 MB](super_resolution_test_image.gz) |  1.5.0  | 10
```
Comment (prasanthpul) [inline vision/super_resolution/super_resolution_cnn_2016/README.md]: "v1" not needed.

onnx file name shouldn't include "tutorial"
